src/Microsoft.AspNet.Mvc.TagHelpers/Internal/AttributeMatcher.cs

```diff
@@ -93,20 +93,26 @@ public static class AttributeMatcher
             var presentAttributes = new List<string>();
             var missingAttributes = new List<string>();
 
-            foreach (var attribute in requiredAttributes)
+            foreach (var requiredAttribute in requiredAttributes)
             {
-                if (!context.AllAttributes.ContainsName(attribute) ||
```

Comment:
Still messy. Also deserves comments about minimized attributes and `attributeName=""` being considered missing. (I don't remember why this works that way.)